### ISSUE TITLE
When validates_uniqness_of is used on primary key column of type uuid validation fails on update

### ISSUE THREAD

**Issue by montas** [rails/rails#23625]
I ran into this on 5.0.0.beta1 but it behaves the same on beta2.

I'm using uuid column type as primary key for table. Update when model doesn't have validation works as expected. If model have validation, update fails with error of uuid not being uniq. I'm not sure if this is because of incorrect table declaration or if this is bug.

Here is failing test

``` ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  gem 'rails', git: 'git@github.com:rails/rails.git', branch: 'v5.0.0.beta2'
  gem 'arel', github: 'rails/arel'
  gem 'rack', github: 'rack/rack'
  gem 'sprockets', github: 'rails/sprockets'
  gem 'sprockets-rails', github: 'rails/sprockets-rails'
  gem 'sass-rails', github: 'rails/sass-rails'
  gem 'pg'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: 'report_db')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :items, force: true, id: false do |t|
    t.uuid :uuid, primary_key: true
    t.string :title
  end
end

class Item < ActiveRecord::Base
end

class ValidatingItem < Item
  validates_uniqueness_of :uuid
end

class BugTest < Minitest::Test
  def test_update
    item = Item.create!(uuid: SecureRandom.uuid, title: 'item1')
    item.update(title: 'item1-title2')

    assert_empty item.errors

    item2 = ValidatingItem.create!(uuid: SecureRandom.uuid, title: 'item2')
    item2.update(title: 'item2-title2')

    assert_empty item2.errors

  end
end
```

and output

```bash
-- create_table(:items, {:force=>true, :id=>false})
D, [2016-02-12T01:55:14.428369 #96236] DEBUG -- :    (4.7ms)  DROP TABLE "items"
D, [2016-02-12T01:55:14.434925 #96236] DEBUG -- :    (6.2ms)  CREATE TABLE "items" ("uuid" uuid PRIMARY KEY, "title" character varying)
   -> 0.0584s
D, [2016-02-12T01:55:14.454753 #96236] DEBUG -- :   ActiveRecord::InternalMetadata Load (0.6ms)  SELECT  "ar_internal_metadata".* FROM "ar_internal_metadata" ORDER BY "ar_internal_metadata"."key" ASC LIMIT $1  [["LIMIT", 1]]
D, [2016-02-12T01:55:14.465138 #96236] DEBUG -- :    (0.1ms)  BEGIN
D, [2016-02-12T01:55:14.466125 #96236] DEBUG -- :    (0.1ms)  COMMIT
Run options: --seed 15933

# Running:

D, [2016-02-12T01:55:14.497080 #96236] DEBUG -- :    (0.1ms)  BEGIN
D, [2016-02-12T01:55:14.498754 #96236] DEBUG -- :   SQL (0.7ms)  INSERT INTO "items" ("uuid", "title") VALUES ($1, $2) RETURNING "uuid"  [["uuid", "18d12e5c-e9b9-4eaa-b325-8d6c480313f3"], ["title", "item1"]]
D, [2016-02-12T01:55:14.499456 #96236] DEBUG -- :    (0.3ms)  COMMIT
D, [2016-02-12T01:55:14.499714 #96236] DEBUG -- :    (0.1ms)  BEGIN
D, [2016-02-12T01:55:14.501272 #96236] DEBUG -- :   SQL (0.5ms)  UPDATE "items" SET "title" = $1 WHERE "items"."uuid" = $2  [["title", "item1-title2"], ["uuid", "18d12e5c-e9b9-4eaa-b325-8d6c480313f3"]]
D, [2016-02-12T01:55:14.501778 #96236] DEBUG -- :    (0.4ms)  COMMIT
D, [2016-02-12T01:55:14.503960 #96236] DEBUG -- :    (0.1ms)  BEGIN
D, [2016-02-12T01:55:14.512227 #96236] DEBUG -- :   ValidatingItem Exists (0.4ms)  SELECT  1 AS one FROM "items" WHERE "items"."uuid" = $1 LIMIT $2  [["uuid", "b0db8b7c-291e-4206-a819-435428d4e7a8"], ["LIMIT", 1]]
D, [2016-02-12T01:55:14.513066 #96236] DEBUG -- :   SQL (0.3ms)  INSERT INTO "items" ("uuid", "title") VALUES ($1, $2) RETURNING "uuid"  [["uuid", "b0db8b7c-291e-4206-a819-435428d4e7a8"], ["title", "item2"]]
D, [2016-02-12T01:55:14.513550 #96236] DEBUG -- :    (0.3ms)  COMMIT
D, [2016-02-12T01:55:14.513746 #96236] DEBUG -- :    (0.1ms)  BEGIN
D, [2016-02-12T01:55:14.515572 #96236] DEBUG -- :   ValidatingItem Exists (1.0ms)  SELECT  1 AS one FROM "items" WHERE "items"."uuid" = $1 LIMIT $2  [["uuid", "b0db8b7c-291e-4206-a819-435428d4e7a8"], ["LIMIT", 1]]
D, [2016-02-12T01:55:14.532580 #96236] DEBUG -- :    (0.2ms)  ROLLBACK
F

Finished in 0.045453s, 22.0008 runs/s, 88.0030 assertions/s.

  1) Failure:
BugTest#test_update [report.rb:51]:
Expected #<ActiveModel::Errors:0x007fc4854b2458 @base=#<ValidatingItem uuid: "b0db8b7c-291e-4206-a819-435428d4e7a8", title: "item2-title2">, @messages={:uuid=>["has already been taken"]}, @details={:uuid=>[{:error=>:taken, :value=>"b0db8b7c-291e-4206-a819-435428d4e7a8"}]}> to be empty.

1 runs, 4 assertions, 1 failures, 0 errors, 0 skips
```

**Comment by maclover7** [comment]
Can confirm this bug. Occurring on master, but not on `4-2-stable`. Will do a `git bisect`.

**Comment by maclover7** [comment]
119b9181ece399c67213543fb5227b82688b536f is the first bad commit.

**Comment by maclover7** [comment]
cc @sgrif - should this be on the 5.0 milestone?

**Comment by sgrif** [comment]
If its new in 5 and seems serious, then yes.

**Comment by maclover7** [comment]
I opened PR #23628 to try and fix this.

**Comment by sgrif** [comment]
Fixed by #23628
